From b81b87127743002101331c458dd77b5a3052b727 Mon Sep 17 00:00:00 2001
From: Wendy Liang <wendy.liang@xilinx.com>
Date: Tue, 28 Jul 2020 22:40:27 -0700
Subject: [PATCH 1478/1852] misc: xilinx-ai-engine: Add partition control flag

commit 0ce28486635a0137485ebb9799b59b85c17c7fe6 from
https://github.com/Xilinx/linux-xlnx.git

Add partition control flag for AI engine partition operation control.
e.g. not reset on release is added to indicate not to reset the partition
when the partition is released.

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/misc/xilinx-ai-engine/ai-engine-dev.c      | 1 +
 drivers/misc/xilinx-ai-engine/ai-engine-internal.h | 3 +++
 drivers/misc/xilinx-ai-engine/ai-engine-reset.c    | 3 +++
 include/uapi/linux/xlnx-ai-engine.h                | 6 ++++++
 4 files changed, 13 insertions(+)

diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-dev.c b/drivers/misc/xilinx-ai-engine/ai-engine-dev.c
index 042c52372bb3..25c2e3cfcacb 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-dev.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-dev.c
@@ -199,6 +199,7 @@ static int aie_partition_get(struct aie_partition *apart,
 	apart->filep = filep;
 
 	apart->status = XAIE_PART_STATUS_INUSE;
+	apart->cntrflag = req->flag;
 	return 0;
 }
 
diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-internal.h b/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
index 5de4badd7c06..7d44d34aa82d 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-internal.h
@@ -176,6 +176,8 @@ struct aie_part_bridge {
  * @partition_id: partition id. Partition ID is the identifier
  *		  of the AI engine partition in the system.
  * @status: indicate if the partition is in use
+ * @cntrflag: partition control flag. e.g. whether to reset columns when
+ *	      the partition is released
  */
 struct aie_partition {
 	struct list_head node;
@@ -188,6 +190,7 @@ struct aie_partition {
 	struct device dev;
 	u32 partition_id;
 	u32 status;
+	u32 cntrflag;
 };
 
 extern struct class *aie_class;
diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-reset.c b/drivers/misc/xilinx-ai-engine/ai-engine-reset.c
index d80b8cc2a145..b67ef5e60e50 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-reset.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-reset.c
@@ -141,6 +141,9 @@ int aie_part_clean(struct aie_partition *apart)
 	struct aie_device *adev = apart->adev;
 	int ret;
 
+	if (apart->cntrflag & XAIE_PART_NOT_RST_ON_RELEASE)
+		return 0;
+
 	aie_part_set_cols_reset(apart, true);
 
 	ret = apart->adev->ops->reset_shim(adev, &apart->range);
diff --git a/include/uapi/linux/xlnx-ai-engine.h b/include/uapi/linux/xlnx-ai-engine.h
index e82d4348fd5f..4f0407d9338c 100644
--- a/include/uapi/linux/xlnx-ai-engine.h
+++ b/include/uapi/linux/xlnx-ai-engine.h
@@ -18,6 +18,12 @@ enum aie_reg_op {
 /* AI engine partition bridge is enabled */
 #define XAIE_PART_STATUS_BRIDGE_ENABLED	(1U << 1)
 
+/*
+ * AI engine partition control flags
+ */
+/* Not reset when release AI engine partition */
+#define XAIE_PART_NOT_RST_ON_RELEASE	0x00000001U
+
 /**
  * struct aie_location - AIE location information
  * @col: column id
-- 
2.31.1

